FF119 Add PublicKeyCredential.parseCreationOptionsFromJSON() (#29436)
mdn/content: files/en-us/web/api/publickeycredential/tojson/index.md

## Syntax

```js-nolint
toJSON()
```

### Parameters

None.

### Return value

A JSON-friendly representation of a [`PublicKeyCredential`](/en-US/docs/Web/API/PublicKeyCredential) object, with included properties that depend on whether the credential was returned by [`navigator.credentials.create()`](/en-US/docs/Web/API/CredentialsContainer/create) on registration or [`navigator.credentials.get()`](/en-US/docs/Web/API/CredentialsContainer/get) when [authenticating a user](/en-US/docs/Web/API/Web_Authentication_API#authenticating_a_user).

The values and types are the same as for [`PublicKeyCredential`](/en-US/docs/Web/API/PublicKeyCredential), with the exception that [base64url](/en-US/docs/Glossary/Base64) encoded strings used in place of buffer properties.

The object properties are:

- `id`
  - : The value returned by {{domxref("PublicKeyCredential.id")}}
- `rawId`

>   - : The value returned by {{domxref("PublicKeyCredential.rawId")}} and then [base64url](/en-US/docs/Glossary/Base64) encoding the returned value.

Author: Yes, I played with this a little, but I much prefer your version.

- `authenticatorAttachment` {{optional_inline}}
  - : The value returned by {{domxref("PublicKeyCredential.authenticatorAttachment")}}.
- `type`
  - : The string `"public-key"`.
- `clientExtensionResults`

>   - : The result of calling {{domxref("PublicKeyCredential.getClientExtensionResults()")}} and [base64url](/en-US/docs/Glossary/Base64) encoding the values of each array entry.

Reviewer: Again, strangely worded. How about 

```suggestion
  - : An array contaning [base64url](/en-US/docs/Glossary/Base64)-encoded versions of the values returned by {{domxref("PublicKeyCredential.getClientExtensionResults()")}}.
```

?

- `response`

  - : The response property object depends on whether the credentials are returned following a registration or authentication operation.

    - When registering a new user `response` will be a JSON-friendly version of {{domxref("AuthenticatorAttestationResponse")}} where buffer values have been [base64url](/en-US/docs/Glossary/Base64) encoded.

    - When authenticating a user the returned value will be a JSON-friendly version of {{domxref("AuthenticatorAssertionResponse")}} where buffer values have been [base64url](/en-US/docs/Glossary/Base64) encoded.

## Examples

When registering a new user the server will supply information about the expected credentials (below shown as the already parsed object `publicKey`).

> First we pass these to [`navigator.credentials.create()`](/en-US/docs/Web/API/CredentialsContainer/create), which returns a promise that eventually resolves to a {{domxref("PublicKeyCredential")}}.

Author: NOte, code below is untested, but I think is what it should look like. Gotta love await/async

```js
const newCredentialInfo = await navigator.credentials.create({ publicKey });
```

The credentials are then posted back to the server as JSON data.

> We call [JSON.stringify()](/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) to serialize the credential object (which in turn calls `toJSON()`).

Reviewer: Ummm, this is really confusing.

So you are serializing the `PublicKeyCredential` object returned by the `create()` call, and then sending that to the server for storage? And then `toJSON()` is called on it sometime after that, magically?

From what the method description says, it sounds like you should be running `PublicKeyCredential.toJSON()`, and then serializing the returned result from that method and sending the result to the server?

Author: Ah, the order is wrong: it was supposed to be "this is what we are doing and then this is how we do it", but it reads as do this then that.

I'll look at this.

Reviewer: Ah, I've read the updated page, and what is there does make more sense now. Thanks.

One thing I noticed — the two linked instances of "JSON.stringify()" in that article are in normal font — they need to be in code font.

---
title: "PublicKeyCredential: toJSON() method"
short-title: toJSON()
slug: Web/API/PublicKeyCredential/toJSON
page-type: web-api-instance-method
status:
  - experimental
browser-compat: api.PublicKeyCredential.toJSON

> ---

Author: @chrisdavidmills Updated, and I think now clear. 
- It returns a "JSON type representation" of the credential with format that depends on if registering or getting credential.
- This is used by stringify and not intended for directly calling.

Reviewer: Yup yup. This works well.
